content/en/docs/concepts/services-networking/network-policies.md

```diff
@@ -99,6 +99,62 @@ So, the example NetworkPolicy:
 
 See the [Declare Network Policy](/docs/tasks/administer-cluster/declare-network-policy/) walkthrough for further examples.
 
+## Behavior of `to` and `from` selectors
+
+There are four kinds of selectors that can be specified in an `ingress` `from` section or `egress` `to` section:
+
+__podSelector__: This selects particular pods in the same namespace as the `NetworkPolicy` which should be allowed as ingress sources or egress destinations.
+
+__namespaceSelector__: This selects particular namespaces for which all pods should be allowed as ingress sources or egress destinations.
+
+__namespaceSelector__ *and* __podSelector__: A single `to`/`from` entry that specifies both `namespaceSelector` and `podSelector` selects particular pods within particular namespaces. Be careful to use correct YAML syntax; this policy:
+
+```yaml
+  ...
+  ingress:
+  - from:
+    - namespaceSelector:
+        matchLabels:
+          user: alice
+      podSelector:
+        matchLabels:
+          role: client
+  ...
+```
+
+contains a single `from` element allowing connections from pods with the label `role=client` in namespaces with the label `user=alice`. But *this* policy:
+
+```yaml
+  ...
+  ingress:
+  - from:
+    - namespaceSelector:
+        matchLabels:
+          user: alice
+    - podSelector:
+        matchLabels:
+          role: client
+  ...
+```
+
+contains two elements in the `from` array, and allows connections from pods in the local namespace with the label `role=client`, *or* from any pod in any namespace with the label `user=alice`.
+
+When in doubt, use `kubectl describe` to see how Kubernetes has interpreted the policy.
+
+__ipBlock__: This selects particular IP addresses to allow as ingress sources or egress destinations. Normally these would be cluster-external IPs, since pod IPs are ephemeral and unpredictable.
+
+In general, `ipBlock` `ingress` rules select traffic based on the source IP address *as seen
+by the selected destination pod*. For traffic coming from outside the cluster, this is
+heavily dependent on the exact details of the network plugin, cloud provider,
+`LoadBalancer`/`Ingress` implementation, etc. With some ingress options you may be able to
+filter based on the actual original source IP. In other cases, the "source IP" that the
+`NetworkPolicy` acts on may be the IP of a `LoadBalancer` or of the pod's node, etc.
+
+Likewise, `ipBlock` `egress` rules generally select traffic based on the destination IP
+provided by the source pod, and in particular, connections to a `Service` IP that then
+gets redirected to a cluster-external IP may not be handled the same way as connections
```

Comment:
I think the ingress section above makes reasonable sense to me. This part is perhaps less obvious to me than the ingress case.

For egress podSelector / namespaceSelector, the enforcement is typically done post service DNAT, and so it's perhaps intuitive that the same would be true for ipBlock as well. I'm not 100% convinced either way.

For both cases, I'm slightly concerned that we may be proscribing a bit too heavily, and a general "Some plugins provide subtly different behaviors, so refer to each plugin's documentation for more information" type statement might be more appropriate.

Reply:
I'm not saying it *should* work this way, but it seems like it probably inevitably *will* work this way for any plugins that don't have their own integrated kube-proxy; they can't enforce the rule before the connection passes through kube-proxy because they don't know what the final destination IP will be, but they probably can't enforce the rule after the connection passes through kube-proxy either, because they can no longer tell what pod the connection came from at that point (and there may not be any hooks left to let the plugin grab the packet again after it has gone through POSTROUTING anyway).

I'm fine with trying to avoid proscribing. That's what the "in general" was about but I can waffle even more than that.

Reply:
I guess the other question is how ipBlock egress rules interact with the service CIDR. I was assuming that if you are allowed to egress to a pod directly, you are also allowed to egress to the pod via any service IP that points to it, without needing to have an ipBlock allowing egress to that service IP or to the service CIDR in general. And in fact, I was assuming that you couldn't block access to service IPs even by doing

    egress:
    - to:
      - ipBlock:
          cidr: 0.0.0.0/0
          except:
          - 172.30.0.0/16

but maybe that's either wrong or implementation-defined?